docs: :pencil2: minor text edits to the first part of `pipelines` (#230)
rostools/r-cubed-advanced: sessions/pipelines.qmd

::: {.callout-note collapse="true"}
## :teacher: Instructor note

Before beginning, encourage learners to recall what they remember from
the previous session, either with something like Mentimeter or verbally.
Preferably with something like Mentimeter because it allows everyone to
participate, not just the ones who are more comfortable being vocal to
the whole group.
:::

> When working on a data analyses, you have probably experienced (many)

Reviewer: ```suggestion
When doing analyses on data, you have probably experienced (many)
```

times where you forget the order in which other code should run or which
parts need to be rerun to update results. Things can get confusing
quickly, even in relatively simple projects. This becomes even more
challenging when you return to a project after a month or two and have
forgotten the state of the analysis and the project as a whole.

That's where formal data analysis pipeline tools come in. By organising
your analysis into distinct steps, with clear inputs and outputs, and

> adding these steps to a pipeline that tracks those them, you can make

Reviewer: ```suggestion
adding these steps to a pipeline that tracks them, you can make
```

However, in research, especially around data collection and analysis, we
often hear or read about "pipelines". But looking closer, these aren't
actual pipelines, because the individual steps are not very clear and

> not well described, and they often requiring a fair amount of manual

Reviewer: ```suggestion
not well described, and they often require a fair amount of manual
```

human attention and intervention. Particularly within computational
environments, a pipeline should be a set of data processing steps
connected in a series with a specific order; the output of one step is
the input to the next. This means that there actually should be minimal
to *no* human intervention from raw input to finished output.

> Why aren't these data "pipelines" in most research actual pipelines?

Reviewer: ```suggestion
Why aren't these computational data pipelines found in most of the "pipelines" described in research?
```

Because:

> 1.  Anything with data ultimately must be on the computer

Reviewer: ```suggestion
1.  Anything with data ultimately must be on the computer,
```

6.  And, research is largely non-reproducible [@Trisovic2022;
    @Seibold2021; @Laurinavichyute2021].

A data analysis pipeline would by definition be a readable and
reproducible data analysis. Unfortunately, we researchers as a group
don't make use of the tools to implement actual data analysis pipelines.

This isn't to diminish the work of researchers, but a basic observation
on the systemic, social, and structural environment surrounding us. We
as researchers are not necessarily trained in writing code, nor do we
have a strong culture and incentive structure around learning, sharing,
reviewing, and improving code. We are also very rarely allowed to get
(or use) funds to hire people who *are* trained and skilled in

> programmatic thinking and coding Otherwise courses like this wouldn't

Reviewer: ```suggestion
programmatic thinking and coding. Otherwise courses like this wouldn't
```

```
:::

Mostly, the`_targets.R` script contains comments and a basic set up to

> get you started. But, notice `tar_target()` function used at the end of

Reviewer: ```suggestion
get you started. But, notice the `tar_target()` function used at the end of
```

model_results <- run_linear_reg(processed_data)
```

With the function-oriented workflow, each function takes an input and
contains all the code to create one result as its output. This could be,
for instance, a figure in a paper.

If you've taken the [intermediate R
course](https://r-cubed-intermediate.rostools.org/), you'll notice that
this function-oriented workflow is the workflow we covered in that
course. There are so many advantages to this type of workflow which is

> why many powerful R packages are designed around making use of this type

Reviewer: ```suggestion
why many powerful R packages are designed around making use of this
```

```{r mean-sd-by-each-metabolite}
#| filename: "doc/learning.qmd"
#| eval: true
# In the {r basic-stats} chunk
lipidomics |>
  group_by(metabolite) |>
  summarise(across(value, list(mean = mean, sd = sd))) |>
  mutate(across(where(is.numeric), ~ round(.x, digits = 1)))
```

After that, style the file using {{< var keybind.styler >}}. Then,

> commir the changes to the Git history with {{< var keybind.git >}}.

Reviewer: ```suggestion
commit the changes to the Git history with {{< var keybind.git >}}.
```

## :technologist: Exercise: Convert summary statistics code into a function

> Time: \~20 minutes.

In the `doc/learning.qmd` file, use the "function-oriented" workflow, as
taught in the [intermediate
course](https://r-cubed-intermediate.rostools.org/sessions/functions#the-basics-of-a-function),
to take the code we wrote above and convert it into a function.

Complete these tasks:

1.  Convert the code into a function by wrapping it with
    `function() {...}` and name the new function `descriptive_stats`.
    Here is some scaffolding to help you get started:

    ``` {.r filename="doc/learning.qmd"}
    descriptive_stats <- function(___) {
      ___
    }
    ```

2.  Replace `lipidomics` in the code we wrote before with `data` and put
    `data` as an argument inside the brackets of `function()`.

3.  At the start of each function we use inside our `descriptive_stats`
    function, add the name of the package like so: `packagename::`.

>     E.g., add `dplyr::` before `group_by()` because it's from from

Reviewer: ```suggestion
    E.g., add `dplyr::` before `group_by()` because it's from
```

>     the`{dplyr}`package. Remember that you can look up which package a

Reviewer: ```suggestion
    the`{dplyr}` package. Remember that you can look up which package a
```

>     function comes from, by writing `?functionname` in the Console.

Reviewer: ```suggestion
    function comes from by writing `?functionname` in the Console.
```

  knitr::kable(caption = "The mean and standard deviation of metabolites in the lipidomics dataset.")
```

> rerun `targets::tar_visnetwork()` using {{< var keybind.targets-vis >}}

Reviewer: ```suggestion
Rerun `targets::tar_visnetwork()` using {{< var keybind.targets-vis >}}
```
